File: deleteDuplicates.py
# https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

"""
The solution to a problem instance is the given list without any duplicate numbers. So, for each number in the given list, include it in the solution iff it isn't a duplicate.

The given list is sorted. In a sorted list, a number is a duplicate iff it's equal to the previous number or the next number. So, for each number in the given list, test whether it's equal to the previous number or the next number. If it is, then skip it; exclude it from the solution. If it isn't, then include it.
"""

class Solution:
	def deleteDuplicates(self, head, prev=None):
		"""
		Include head in the solution or not?
		Include head iff head isn't a duplicate
		head is a duplicate iff its previous node or its next node has the same value
		"""
		if not head:
			return None
		elif (prev and prev.val == head.val) or (head.next and head.next.val == head.val):
			return self.deleteDuplicates(head.next, head)
		else:
			head.next = self.deleteDuplicates(head.next, head)
			return head

File: deleteDuplicates.java
// https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

/*
 * I developed the algorithm in Python (see deleteDuplicates.py)
 * I translated the Python to Java to show prospective employers that I can write Java
 */

class Solution {
	public ListNode deleteDuplicates(ListNode head) {
		return deleteDuplicates(head, null);
	}

	public ListNode deleteDuplicates(ListNode head, ListNode prev) {
		if (head == null)
			return null;
		else if ((prev != null && prev.val == head.val) || (head.next != null && head.next.val == head.val))
			return deleteDuplicates(head.next, head);
		else {
			head.next = deleteDuplicates(head.next, head);
			return head;
		}
	}
}


/*
 * After I solve a problem, I like to study alternative solutions
 * The following program embodies a nifty way to approach the problem
 */

public ListNode deleteDuplicates(ListNode head) {
        if(head==null) return null;
        ListNode FakeHead=new ListNode(0);
        FakeHead.next=head;
        ListNode pre=FakeHead;
        ListNode cur=head;
        while(cur!=null){
            while(cur.next!=null&&cur.val==cur.next.val){
                cur=cur.next;
            }
            if(pre.next==cur){
                pre=pre.next;
            }
            else{
                pre.next=cur.next;
            }
            cur=cur.next;
        }
        return FakeHead.next;
    }
